sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/SchedulerBuilder.java

```diff
@@ -0,0 +1,503 @@
+            setStateStore(new StateStore(persister));
+        }
+        return stateStoreOptional.get();
+    }
+
+    /**
+     * Specifies a custom {@link ConfigStore}.
+     *
+     * The config store persists a copy of the current configuration ('target' configuration),
+     * while also storing historical configurations.
+     */
+    public SchedulerBuilder setConfigStore(ConfigStore<ServiceSpec> configStore) {
+        if (configStoreOptional.isPresent()) {
+            // Any customization of the state store must be applied BEFORE getConfigStore() is ever called.
+            throw new IllegalStateException(
+                    "Config store is already set. Was getConfigStore() invoked before this?");
+        }
+        this.configStoreOptional = Optional.ofNullable(configStore);
+        return this;
+    }
+
+    /**
+     * Returns the {@link ConfigStore} provided via {@link #setConfigStore(ConfigStore)}, or a reasonable default
+     * created via {@link DefaultScheduler#createConfigStore(ServiceSpec, Collection)}.
+     *
+     * In order to avoid cohesiveness issues between this setting and the {@link #build()} step,
+     * {@link #setConfigStore(ConfigStore)} may not be invoked after this has been called.
+     */
+    public ConfigStore<ServiceSpec> getConfigStore() {
+        if (!configStoreOptional.isPresent()) {
+            try {
+                setConfigStore(createConfigStore(serviceSpec, Collections.emptyList(), persister));
+            } catch (ConfigStoreException e) {
+                throw new IllegalStateException("Failed to create default config store", e);
+            }
+        }
+        return configStoreOptional.get();
+    }
+
+    /**
+     * Specifies custom endpoint resources which should be exposed through the scheduler's API server, in addition
+     * to the defaults.
+     */
+    public SchedulerBuilder setCustomResources(Collection<Object> customResources) {
+        this.customResources = customResources;
+        return this;
+    }
+
+    /**
+     * Specifies a custom list of configuration validators to be run when updating to a new target configuration,
+     * or otherwise uses the default validators returned by {@link DefaultScheduler#defaultConfigValidators()}.
+     */
+    public SchedulerBuilder setCustomConfigValidators(Collection<ConfigValidator<ServiceSpec>> customConfigValidators) {
+        this.customConfigValidators = customConfigValidators;
+        return this;
+    }
+
+    /**
+     * Specifies a custom {@link EndpointProducer} to be added to the /endpoints API. This may be used by services
+     * which wish to expose custom endpoint information via that API.
+     *
+     * @param name the name of the endpoint to be exposed
+     * @param endpointProducer the producer to be invoked when the provided endpoint name is requested
+     */
+    public SchedulerBuilder setEndpointProducer(String name, EndpointProducer endpointProducer) {
```

Comment:
This is using `set` because provided producer is indexed against the name. `add` would imply that two consecutive additions against the same name would result in two values, rather than the latter one overridding the prior one.

Reply:
(yes, I was thinking that in this case since the arguments are effectively key-value pairs, `set` is more applicable).